src/main/src/main/java/org/geoserver/security/impl/DefaultResourceAccessManager.java

```diff
@@ -423,5 +657,95 @@ private boolean canAccess(Authentication user, SecureTreeNode node) {
             return access;
         }
     }
+    
+    private SecureTreeNode getNodeForGroup(LayerGroupInfo lg) {
+        SecureTreeNode node;
+        if(lg.getWorkspace() == null) {
+            node = root.getNode(lg.getName());
+        } else {
+            String[] path = getLayerGroupPath(lg);
+            node = root.getNode(path);
+        }
+        return node;
+    }
+    
+    /**
+     * This listener keeps the "layer group" flags in the authorization tree current, in order
+     * to optimize the application of layer group containment rules
+     */
+    class CatalogChangeListener implements CatalogListener {
+
+        @Override
+        public void handleAddEvent(CatalogAddEvent event) throws CatalogException {
+            if(event.getSource() instanceof LayerGroupInfo) {
+                LayerGroupInfo lg = (LayerGroupInfo) event.getSource();
+                addGroupInfo(lg);
+            } 
+        }
+
+        private void addGroupInfo(LayerGroupInfo lg) {
+            SecureTreeNode node = getNodeForGroup(lg);
+            if(node != null) {
+                updateContainedLayerIds(node, lg);
+            }
+        }
+
+        @Override
+        public void handleRemoveEvent(CatalogRemoveEvent event) throws CatalogException {
+            if(event.getSource() instanceof LayerGroupInfo) {
+                LayerGroupInfo lg = (LayerGroupInfo) event.getSource();
+                clearGroupInfo(lg);
+            }
+        }
+
+        @Override
+        public void handleModifyEvent(CatalogModifyEvent event) throws CatalogException {
+            final CatalogInfo source = event.getSource();
+            if(source instanceof LayerGroupInfo) {
+                LayerGroupInfo lg = (LayerGroupInfo) event.getSource();
+                // was the layer group renamed or moved?
+                int nameIdx = event.getPropertyNames().indexOf("name");
+                int wsIdx = event.getPropertyNames().indexOf("workspace");
+                if(nameIdx != -1 || wsIdx != -1) {
+                    // group has been modified, clear the old node (the new one will be updated in ppost modify)
+                    clearGroupInfo(lg);
```

Comment:
I made the following test, created a global layer group and defined a rule for him. Then I updated the layer group adding him to an workspace. The created rule will still exists and reference the global layer group, not sure what should be done here.